File: web/static/fiz/AutocompleteFormElement.js
/* AutocompleteFormElements.js --
 *
 * This file provides Javascript functions needed to implement the
 * AutocompleteFormElement class.  One Fiz.AutocompleteFormElement
 * is created for each AutocompleteFormElement Java object. 
 * Methods on the object are invoked for functions such as
 * populating the autcomplete list, fetching data from the server,
 * and navigating the list of results.
 *
 * Copyright (c) 2009 Stanford University
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

// The following lines are used by Fiz to manage Javascript dependencies.
// Fiz:include static/fiz/Fiz.js
// Fiz:include static/fiz/Ajax.js

// Define symbolic names for keycodes used by the AutocompleteFormElement
Fiz.Key = {
    UP: 38, DOWN: 40, ENTER: 13, ESC: 27, TAB: 9
};

/**
 * Create a AutocompleteFormElement object.
 * @param id                       Id for the {@code <div>} element
 *                                 that represents the autcomplete form.
 */
Fiz.AutocompleteFormElement = function(id)
{
    // Set the id identifying the <input> element for this form
    // element and from which all related ids are derived
    this.id = id;

    // Indicates whether the dropdown is visible or not
    this.isOpen = false;
    
    // Used to disable to onblur event handler. We want to keep the dropdown 
    // open if the user is clicking a choice in the dropdown. By default,
    // the input field will lose focus, thus closing the dropdown before the
    // user's click is registered.
    this.keepOpen = false;
        
    // The currently selected choice in the dropdown menu (this value will be
    // shown in the input field)
    this.selected = null;
    
    // The currently highlighted choice in the dropdown menu (this value may not
    // neccessarily be displayed in the input field if the choice is highlighted
    // when the mouse hovers over it)
    this.highlighted = null;
    
    // <input> element for which this autocomplete object is registered
    this.input = document.getElementById(id);
    
    // Stores the last value the user entered into the <input> field so that
    // it can be restored if needed
    this.lastUserInput = this.input.value;
    
    // Stores the last value displayed in the <input> field to prevent
    // refreshMenu from being called if the value has not changed
    this.lastInputValue = this.input.value;
    
    // Disables browser built-in autocomplete, not supported in all browsers and
    // not part of the web standard
    this.input.setAttribute('autocomplete', 'off');
    
    // Dropdown menu of autocomplete choices
    this.dropdown = document.getElementById(id + '_dropdown');
}

/**
 * This method is invoked when a keydown event occurs while the <input> element
 * is in focus. Overrides the default key behavior of the tab, enter, esc,
 * up arrow, down arrow keys in order to provide mechanisms such as keyboard
 * navigation and selection.
 * @param e                    (Event) The event object containing information
 *                             about the keystroke and event type.
 */
Fiz.AutocompleteFormElement.prototype.captureKeydown = function(e)
{
    if (this.isOpen) {
        switch(Fiz.getKeyCode(e)) {
            // Move the user selection up the list of choices
            case Fiz.Key.UP:
                this.prevResult();
                // Prevent the cursor from moving to the beginning of the input
                // field when the user is navigating
                e.preventDefault();
                return false;
            // Move the user selection down the list of choices
            case Fiz.Key.DOWN:
                this.nextResult();
                // Prevent the cursor from moving to the end of the input field
                // when the user is navigating
                e.preventDefault();
                return false;
            // Finalize the user selection and put the value of the selection in
            // the input field
            case Fiz.Key.ENTER:
            // Manually hide the dropdown
            case Fiz.Key.ESC:
                this.hideDropdown(true);
                e.preventDefault();
                return false;
            // Hide the dropdown when the user navigats to another field
            case Fiz.Key.TAB:
                this.hideDropdown(true);
                return true;
        }
    } else {
        // Open the dropdown again if the user wants to pick another choice from
        // the autocomplete list
        if (Fiz.getKeyCode(e) == Fiz.Key.UP
                || Fiz.getKeyCode(e) == Fiz.Key.DOWN) {
            this.showDropdown();
            e.preventDefault();
            return false;
        }
    }    
}

/**
 * This method opens up the autocomplete dropdown. If there is
 * currently no query, this method does nothing.
 */
Fiz.AutocompleteFormElement.prototype.showDropdown = function()
{
    if (this.input.value == '') {
        this.hideDropdown();
        return;
    }
    
    // Reset the state of the dropdown menu
    this.isOpen = true;
    
    // Find the location of the input field
    var x = 0;
    var y = this.input.offsetHeight - 1;
        
    // Attach dropdown to the input field
    this.dropdown.style.top = y + 'px';
    this.dropdown.style.left = x + 'px';
    this.dropdown.style.display = 'block';
}

/**
 * This method hides the autocomplete dropdown if hiding is enabled (hiding
 * will be disabled when the user mouses over dropdown to select something).
 * @param forceHide            (Boolean) Forces the dropdown to hide even if
 *                             the user has their mouse over the dropdown
 */
Fiz.AutocompleteFormElement.prototype.hideDropdown = function(forceHide)
{
    if (forceHide) {
        this.keepOpen = false;
    }
    if (!this.keepOpen) {
        this.isOpen = false;
        this.dropdown.style.display = 'none';
        this.input.focus();
    }
}

/**
 * This method selects the autocomplete entry, highlighting it and filling
 * in the input field.
 * @param choice              (Element) An <li> element in the dropdown menu
 * @param done                (Boolean) Hides the dropdown after selection
 *                            if true
 */
Fiz.AutocompleteFormElement.prototype.selectChoice = function(choice, done)
{    
    // Select a new option
    if (null == choice) {
        this.input.value = this.lastUserInput;
    } else {
        this.input.value = Fiz.getText(choice);        
    }
    
    this.lastInputValue = this.input.value;
    this.selected = choice;

    // Even if done is set to true (the dropdown closes after the selection),
    // we highlight the option in case the dropdown is opened again.
    this.highlightChoice(choice);
    
    if (done) {
        this.hideDropdown(true);
    }
}

/**
 * This method highlights the user's current choice in the dropdown.
 * @param choice                    (Element) The element to highlight
 */
Fiz.AutocompleteFormElement.prototype.highlightChoice = function(choice)
{
    // Clear the old highlighted element
    if (null != this.highlighted) {
        Fiz.removeClass(this.highlighted, 'highlight');
    }
    
    // Highlight a new option
    if (null != choice) {
        Fiz.addClass(choice, 'highlight');
    }
    this.highlighted = choice;
}

/**
 * This method moves the selection up on the dropdown list.
 */
Fiz.AutocompleteFormElement.prototype.prevResult = function()
{
    if (null == this.highlighted) {
        // The user is currently "focused" on the input field so
        // pressing up starts from the bottom of the autocomplete list
        this.selectChoice(document
        		.getElementById(this.id + "_choices").lastChild);
    } else {
        this.selectChoice(this.highlighted.previousSibling);
    }
}

/**
 * This method moves the selection down on the dropdown list.
 */
Fiz.AutocompleteFormElement.prototype.nextResult = function()
{
    if (null == this.highlighted) {
        // The user is currently "focused" on the input field so
        // pressing up starts from the bottom of the autocomplete list
        this.selectChoice(document
        		.getElementById(this.id + "_choices").firstChild);
    } else {
        this.selectChoice(this.highlighted.nextSibling);
    }
}

/**
 * This method generates an Ajax call that retrieves the autocomplete
 * results from the server.
 * @return                    (Fiz.Ajax) Returns the Ajax object for testing
 *                            purposes
 */
Fiz.AutocompleteFormElement.prototype.refreshMenu = function()
{
    if (this.input.value == '') {
        this.hideDropdown(true);
        return null;
    } else {
        // We launch a new Ajax request if the form value has changed.
        if (this.input.value != this.lastInputValue) {
            // Reset the state of the dropdown since we are clearing all its
            // contents
            this.selected = null;
            this.highlighted = null;
            this.lastInputValue = this.input.value;
            this.lastUserInput = this.input.value;
            return new Fiz.Ajax({
                    url: '/AutocompleteFormElement/ajaxQuery',
                    data: { id: this.id, userInput: this.input.value }
            });
        }
        return null;
    }
}

File: src/org/fiz/section/AutocompleteFormElement.java
/* Copyright (c) 2009 Stanford University
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

package org.fiz.section;

import org.fiz.*;
import java.io.Serializable;

/**
 * The AutocompleteFormElement helps users fill in a field by providing
 * suggestions or options for completing what the user has already typed. It
 * supports the following properties (additional properties applicable to all
 * form elements may be found in {@link FormElement}):
 *   dataFactory:       (required) Identifies a factory method that takes a
 *                      single String argument and returns a dataset that
 *                      contains possible autocomplete choices.
 *                      Must have the form {@code class.method}, where
 *                      {@code class} is a Java class name and {@code method}
 *                      is a static method in the class.  The values expected in
 *                      the dataset are described below.  The string
 *                      argument to the factory method is current user input
 *                      for which we are trying to autocomplete.
 *   choiceName:        (optional) Identifies the field in each record returned
 *                      by the Ajax which contains the value to be used for
 *                      the autocompletion choice. Defaults to "choice."
 *
 * The dataset generated from {@code dataFactory} consists of one {@code record}
 * child for each autocomplete choice. Each autocomplete choice will be a
 * dataset containing one key-value pair where the key is specified by the
 * {@code choiceName} parameter and the value will be shown in the autocomplete
 * dropdown.
 *
 * AutocompleteFormElement automatically sets the following {@code class}
 * attributes for use in CSS:
 *   dropdown:       The {@code <div>} containing the autocomplete dropdown.
 *   highlight:      The {@code <li>} containing the currently highlighted
 *                   choice in the dropdown.
 */
public class AutocompleteFormElement extends FormElement implements DirectAjax {
    // One object of the following class is stored as a page property
    // for each AutocompleteFormElement in a page: it holds data that we
    // will need later on to process Ajax requests to request autocomplete
    // information for the form field.
    protected static class PageProperty implements Serializable {
        // The following variables are just copies of configuration properties
        // for the section, or null if no such property.
        protected String id;
        protected String dataFactory;
        protected String choiceName;

        public PageProperty(String id, String dataFactory, String choiceName) {
            this.id = id;
            this.dataFactory = dataFactory;
            this.choiceName = choiceName;
        }
    }

    // Reference to the page state for this section, stored as a page
    // property named {@code AutocompleteFormElement-id}, where {@code id}
    // is the id attribute for the section..
    protected PageProperty pageProperty;

    /**
     * Construct an AutocompleteFormElement from a set of properties that define
     * its configuration.
     *
     * @param properties        Dataset whose values are used to configure the
     *                          element. See the class documentation above for
     *                          a list of supported values.
     */
    public AutocompleteFormElement(Dataset properties) {
        super(properties);
        String choiceName = properties.checkString("choiceName");
        if (choiceName == null) {
            choiceName = "choice";
        }
        pageProperty = new PageProperty(properties.getString("id"),
                properties.getString("dataFactory"), choiceName);
    }

    /**
     * This method is an Ajax entry point, invoked to fill in the menu
     * containing candidate completions of the current text.
     * @param cr                    Overall information about the client
     *                              request being serviced; there must
     *                              be a {@code id} value, which identifies the
     *                              AutocompleteFormElement, and a
     *                              {@code userInput} value, which is the value
     *                              for which we are trying to autocomplete, in
     *                              the main dataset.
     */
    public static void ajaxQuery(ClientRequest cr) {
        Dataset main = cr.getMainDataset();
        PageProperty pageProperty = (PageProperty)
                cr.getPageProperty(main.getString("id"));

        String query = main.getString("userInput");

        Dataset data = (Dataset) Util.invokeStaticMethod(
                pageProperty.dataFactory, query);

        if (data.getDatasetList("record").size() > 0) {
            StringBuilder html = new StringBuilder();
            Template.appendHtml(html, "<ul id=\"@(1)_choices\">",
                    pageProperty.id);
            for(Dataset dataValue : data.getDatasetList("record")) {
                String value = dataValue.getString(pageProperty.choiceName);

                // Extracts the position of the query to highlight it
                int queryIndex = value.toLowerCase()
                        .indexOf(query.toLowerCase());
                String js1 = Template.expandJs(
                        "Fiz.ids.@1.selectChoice(this, true)", pageProperty.id);
                String js2 = Template.expandJs(
                        "Fiz.ids.@1.highlightChoice(this)", pageProperty.id);

                if (queryIndex == -1) {
                    Template.appendHtml(html, "<li onclick=\"@1\" " +
                            "onmouseover=\"@2\">@3</li>", js1, js2,value
                    );
                } else {
                    Template.appendHtml(html, "<li onclick=\"@1\" " +
                            "onmouseover=\"@2\">@3<strong>@4</strong>@5</li>",
                            js1, js2,
                            value.substring(0, queryIndex),
                            value.substring(queryIndex, query.length()),
                            value.substring(queryIndex + query.length(),
                                    value.length())
                    );
                }
            }
            html.append("</ul>");
            cr.updateElement(pageProperty.id + "_dropdown", html.toString());
            cr.evalJavascript("Fiz.ids.@1.showDropdown();\n",
                    pageProperty.id);
        } else {
            cr.updateElement(pageProperty.id + "_dropdown", "");
            cr.evalJavascript("Fiz.ids.@1.hideDropdown(true);\n",
                    pageProperty.id);
        }

    }

    /**
     * This method is invoked to generate HTML for this form element.
     *
     * @param cr                Overall information about the client
     *                          request being serviced.
     * @param data              Data for the form (a CompoundDataset
     *                          including both form data, if any, and
     *                          the main dataset).
     */
    @Override
    public void render(ClientRequest cr, Dataset data) {
        StringBuilder out = cr.getHtml().getBody();
        cr.setPageProperty(pageProperty.id, pageProperty);
        cr.setAuthToken();

        Template.appendHtml(out,
                "\n<!-- Start AutocompleteFormElement @id -->\n" +
                "<div class=\"@class?{AutocompleteFormElement}\" " +
                "id=\"@(id)_container\">\n", properties);
        Template.appendHtml(out,
                "  <input type=\"text\" id=\"@id\" name=\"@id\" " +
                "onkeyup=\"Fiz.ids.@id.refreshMenu()\" " +
                "onkeydown=\"Fiz.ids.@id.captureKeydown(event)\" " +
                "onblur=\"Fiz.ids.@id.hideDropdown()\" " +
                "{{value=\"@1\"}} />\n" +
                "  <div id=\"@(id)_dropdown\" class=\"dropdown\" " +
                "onmouseover=\"Fiz.ids.@id.keepOpen = true\" " +
                "onmouseout=\"Fiz.ids.@id.keepOpen = false\"></div>\n",
                properties, data.checkString(id));
        out.append("</div>\n");
        Template.appendHtml(out, "<!-- End AutocompleteFormElement @id -->\n",
                properties);

        // Generate a Javascript object containing information about the form.
        cr.evalJavascript("Fiz.ids.@id = new Fiz.AutocompleteFormElement(" +
                "\"@id\");\n", properties);
        cr.getHtml().includeCssFile("AutocompleteFormElement.css");
        cr.getHtml().includeJsFile("static/fiz/AutocompleteFormElement.js");
    }
}
